lib/pbio/drv/usb/usb_ev3.c

```diff
@@ -52,500 +28,823 @@
+    for (;;) {
+        PBIO_OS_AWAIT_UNTIL(state, usb_rx_is_ready);
 
-    // Enable the clocks to the USB and PHY modules. Inspired by EV3RT.
-    HWREG(CFGCHIP2_USBPHYCTRL) &= ~SYSCFG_CFGCHIP2_USB0OTGMODE;
-    HWREG(CFGCHIP2_USBPHYCTRL) |= CFGCHIP2_FORCE_DEVICE;  // Force USB device operation
-    HWREG(CFGCHIP2_USBPHYCTRL) |= CFGCHIP2_REFFREQ_24MHZ; // 24 MHz OSCIN
+        if (usb_rx_is_ready) {
+            __asm__ volatile ("" ::: "memory");
 
-    process_start(&pbdrv_usb_process);
+            uint32_t usb_rx_sz = cppi_descriptors[CPPI_DESC_RX].word0.pktLength;
 
-    /*
-    ** Registers the UARTIsr in the Interrupt Vector Table of AINTC.
-    ** The event number of UART2 interrupt is 61.
-    */
+            // TODO: Remove this echo test
+            unsigned int i;
+            for (i = 0; i < usb_rx_sz; i++) {
+                ep1_tx_response_buf[i] = ep1_rx_buf[i] + 1;
+            }
+            for (; i < 512; i++) {
+                ep1_tx_response_buf[i] = 0xaa;
+            }
 
-    IntRegister(SYS_INT_USB0, USB0DeviceIntHandler);
+            (void)ep1_tx_status_buf;
+            (void)ep1_tx_stdout_buf;
 
-    /*
-    ** Map the channel number 2 of AINTC to system interrupt 61.
-    ** Channel number 2 of AINTC is mapped to IRQ interrupt of ARM9 processor.
-    */
-    IntChannelSet(SYS_INT_USB0, 2);
+            unsigned int tx_sz = is_usb_hs ? PYBRICKS_EP_PKT_SZ_HS : PYBRICKS_EP_PKT_SZ_FS;
+            if (!usb_tx_response_is_not_ready) {
+                usb_tx_response_is_not_ready = 1;
+                usb_setup_tx_dma_desc(CPPI_DESC_TX_RESPONSE, ep1_tx_response_buf, tx_sz);
+            }
 
-    /* Enable the system interrupt number 61 in AINTC.*/
-    IntSystemEnable(SYS_INT_USB0);
+            // Re-queue RX buffer after processing is complete
+            usb_rx_is_ready = 0;
+            usb_setup_rx_dma_desc();
+        }
+    }
 
-    //
-    // Not configured initially.
-    //
-    g_bUSBConfigured = false;
+    PBIO_OS_ASYNC_END(PBIO_SUCCESS);
+}
 
-    //
-    // Initialize the Rx and TX Buffers
-    //
-    USBBufferInit((tUSBBuffer *)&g_sTxBuffer);
-    USBBufferInit((tUSBBuffer *)&g_sRxBuffer);
+void pbdrv_usb_init(void) {
+    // This reset sequence is from Example 34-1 in the AM1808 TRM (spruh82c.pdf)
+    // Because PHYs and clocking are... as they tend to be, use the precise sequence
+    // of operations specified.
 
-    USBBufferFlush(&g_sTxBuffer);
-    USBBufferFlush(&g_sRxBuffer);
+    // Power on and reset the controller
+    PSCModuleControl(SOC_PSC_1_REGS, HW_PSC_USB0, PSC_POWERDOMAIN_ALWAYS_ON, PSC_MDCTL_NEXT_ENABLE);
+    USBReset(USB0_BASE);
 
-    g_sCDCSerDeviceInfo.sCallbacks.pfnSuspendHandler = g_sCDCSerDeviceInfo.sCallbacks.pfnDisconnectHandler;
-    //
-    // Pass our device information to the USB library and place the device
-    // on the bus.
+    // Reset the PHY
+    HWREG(CFGCHIP2_USBPHYCTRL) |= CFGCHIP2_RESET;
+    for (int i = 0; i < 50; i++) {__asm__ volatile ("");
+    }
```

Comment:
```suggestion
    for (int i = 0; i < 50; i++) {
        __asm__ volatile (""); // this is a delay loop, don't optimize it out
    }
```

Or could use `pbdrv_clock_busy_delay_us();`

And would be helpful to have a comment explaining why this delay is needed. I'm guessing because "that is how TI did it"?